File: python/fft/fft_10.py
# Welcome to this delightful program that will take you on a journey through the magical world of Fast Fourier Transform (FFT).
# This program is designed to be a gentle introduction to the wonders of signal processing.
# Along the way, we'll encounter various charming functions and variables that will help us achieve our goal.
# So, let's embark on this enchanting adventure together!

import numpy as np

def frodo_fft(input_signal):
    # This function will be our brave hero, Frodo, who will carry the input signal to the land of FFT.
    # Along the way, Frodo will encounter various challenges and helpers.
    samwise = np.fft.fft(input_signal)
    return samwise

def aragorn_input():
    # Aragorn, the ranger, will guide us through the process of gathering user input.
    # He is wise and experienced, and will ensure we get the right input for our journey.
    input_string = input("Please enter the signal values separated by commas: ")
    input_list = input_string.split(',')
    input_signal = [float(value) for value in input_list]
    return input_signal

def legolas_output(output_signal):
    # Legolas, the elf, will gracefully display the output of our FFT journey.
    # His keen eyes will ensure that we see the results clearly.
    print("The FFT of the input signal is:")
    print(output_signal)

def gandalf_main():
    # Gandalf, the wise wizard, will oversee the entire process and ensure everything goes smoothly.
    # He will call upon Frodo, Aragorn, and Legolas to complete the task.
    input_signal = aragorn_input()
    output_signal = frodo_fft(input_signal)
    legolas_output(output_signal)

if __name__ == "__main__":
    # This is where our adventure begins. Gandalf will take charge and lead us through the journey.
    gandalf_main()

File: swift/fft/fft_10.swift
// Welcome to the FFT program, a delightful journey into the world of Fast Fourier Transform.
// This program is designed to help you understand the beauty of signal processing and the magic of frequency domain analysis.
// Let's embark on this enchanting adventure together!

import Foundation

// Function to perform FFT on an array of complex numbers
func fft(_ inputArray: [Complex]) -> [Complex] {
    let n = inputArray.count
    guard n > 1 else { return inputArray }
    
    let evenArray = fft(Array(inputArray.enumerated().filter { $0.offset % 2 == 0 }.map { $0.element }))
    let oddArray = fft(Array(inputArray.enumerated().filter { $0.offset % 2 != 0 }.map { $0.element }))
    
    var result = Array(repeating: Complex(real: 0, imag: 0), count: n)
    let angle = -2.0 * Double.pi / Double(n)
    let w = Complex(real: cos(angle), imag: sin(angle))
    var wk = Complex(real: 1, imag: 0)
    
    for k in 0..<n/2 {
        result[k] = evenArray[k] + wk * oddArray[k]
        result[k + n/2] = evenArray[k] - wk * oddArray[k]
        wk = wk * w
    }
    
    return result
}

// Complex number class
class Complex {
    var real: Double
    var imag: Double
    
    init(real: Double, imag: Double) {
        self.real = real
        self.imag = imag
    }
    
    static func +(lhs: Complex, rhs: Complex) -> Complex {
        return Complex(real: lhs.real + rhs.real, imag: lhs.imag + rhs.imag)
    }
    
    static func -(lhs: Complex, rhs: Complex) -> Complex {
        return Complex(real: lhs.real - rhs.real, imag: lhs.imag - rhs.imag)
    }
    
    static func *(lhs: Complex, rhs: Complex) -> Complex {
        return Complex(real: lhs.real * rhs.real - lhs.imag * rhs.imag, imag: lhs.real * rhs.imag + lhs.imag * rhs.real)
    }
}

// Function to convert a string to a complex number array
func stringToComplexArray(_ input: String) -> [Complex] {
    let components = input.split(separator: " ").map { Double($0) ?? 0.0 }
    var complexArray: [Complex] = []
    for i in stride(from: 0, to: components.count, by: 2) {
        let real = components[i]
        let imag = i + 1 < components.count ? components[i + 1] : 0.0
        complexArray.append(Complex(real: real, imag: imag))
    }
    return complexArray
}

// Function to print complex number array
func printComplexArray(_ array: [Complex]) {
    for complex in array {
        print("(\(complex.real), \(complex.imag))")
    }
}
